### ISSUE TITLE
Github Pre-Release Trigger failure

### ISSUE THREAD

**Issue by ldraney** [department-of-veterans-affairs/notification-api#1214]
## Description
The "pre-release" trigger for Github Actions is not working as expected. However, we discovered that changing the trigger to "published" in our `.github/workflows/release_trigger.yml` file should start triggering the workflow. The problem with the current behavior is that when deploying and creating a pre-release, it doesn't trigger a deploy for staging, which requires manual work.

- [x] discuss with QA

## Steps to Reproduce
1. Create a pre-release
2. Publish the pre-release
3. Observe that the "deploy to staging" workflow is not triggered

## Workaround
Change the trigger in the `.github/workflows/release_trigger.yml` file from `[release, pre-release]` to `[published]`.

## Impact/Urgency
This bug has a medium impact on our system and team, as it causes delays and requires a manual step in our process.

## Expected Behavior
When a pre-release is created, the "deploy to staging" workflow should be triggered automatically.

- [x] The "deploy to staging" workflow should run automatically
- [x] An engineer should not have to manually deploy to staging
- [ ] If this does not fix it, create a new ticket

## QA Considerations
- [x] create a draft pre-release for staging
- [x] after the draft has been saved, publish the pre-release to staging
- [ ] If the code is deployed, the fix works, else we need a new approach.

## Additional Info & Resources
- Link to the `.github/workflows/release_trigger.yml` file where the trigger can be modified: https://github.com/department-of-veterans-affairs/notification-api/blob/master/.github/workflows/release_trigger.yml#L4
- The [Github Actions Documentation](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release) that indicates the change that will likely fix the issue

**Comment by mjones-oddball** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VA-Notify-620d21369d810a00146ed9c8/issues/department-of-veterans-affairs/notification-api/1214?planning-poker) with Zenhub @cris-oddball @EvanParish @jakehova @justaskdavidb2 @k-macmillan @kalbfled @ldraney

**Comment by mjones-oddball** [comment]
@cris-oddball Is this still happening/intermittent? It's been in next up for a while but we previously weren't ready to prioritize it due to the flakiness of the issue and insufficient data. Should it be in the backlog?

**Comment by cris-oddball** [comment]
@mjones-oddball It's intermittent. We should probably work it, since the change is minor and it would be good to know if the fix actually works.

**Comment by cris-oddball** [comment]
![Screen Shot 2023-06-20 at 1.10.57 PM.jpg](https://images.zenhubusercontent.com/630696b479f5cca51e546872/f69ebc9e-dfb9-4e26-b2ff-d23ce82b3a50)

Created a tag from master, after I merged Lucas's code into master that contained the potential fix for the -pre-release trigger. 
The Action did kick off - so in that sense it 'worked' but the deployment failed as shown above.

Link to failed deployment:
https://github.com/department-of-veterans-affairs/notification-api/actions/runs/5326447256

**Comment by ldraney** [comment]
specifically it's failing because of:
![image.png](https://images.zenhubusercontent.com/6361638d438b9c8e844e2b23/a29bb783-9610-4e62-a92c-eb0621d077af)

The task definition has an invalid path.

**Comment by ldraney** [comment]
the problem is this:
```
          script: |
            if ('${{ github.event.action }}' === 'prereleased') {
              core.setOutput('environment', 'staging')
            }
            if ('${{ github.event.action }}' === 'released') {
              core.setOutput('environment', 'prod')
            }
```
needs to be changed to something like this:
```
          script: |
            if (context.payload.release.prerelease) {
              core.setOutput('environment', 'staging')
            } else {
              core.setOutput('environment', 'prod')
            }
```

But the problem with the above is it will deploy to prod every time staging isn't correctly called!  Prod should not be the default. 

I'm considering something like:
```
          script: |
            if (context.payload.release.prerelease) {
              core.setOutput('environment', 'staging')
            } else if (context.payload.release) {
              core.setOutput('environment', 'prod')
            } else {
              core.setOutput('environment', '')
            }
```
But I first need to confirm payload.release will work, but I'm having a hard time finding the relevant documentation.

**Comment by ldraney** [comment]
I'm reading [this documentation](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context) to try and make sure context.payload.release will work

**Comment by ldraney** [comment]
Also perhaps a relevant [stackoverflow](https://stackoverflow.com/questions/70104600/complete-list-of-github-actions-contexts)

**Comment by ldraney** [comment]
Also this [webhook events ](https://docs.github.com/en/webhooks-and-events/webhooks/webhook-events-and-payloads#webhook-payload-example-33) is probably also relevant

**Comment by ldraney** [comment]
I should be able to make a gh cli command to verify this, but this is taking me time to figure out

**Comment by ldraney** [comment]
At first I thought Github context is probably not what I’m looking for because the trigger is an action, not a previous workflow from a push event. A push event occurs when a commit or tag is pushed. 
However, “this object is identical to the webhook payload of the event that triggered the workflow run, and is different for each event. The webhooks for each GitHub Actions event is linked in "Events that trigger workflows." For example, for a workflow run triggered by the push event, this object contains the contents of the push webhook payload.

Note: The prereleased type will not trigger for pre-releases published from draft releases, but the published type will trigger. If you want a workflow to run when stable and pre-releases publish, subscribe to published instead of released and prereleased.

**Comment by ldraney** [comment]
I realized the best way to test triggers is on a personal repo, so I set up some actions similar to notification-api:
![image.png](https://images.zenhubusercontent.com/6361638d438b9c8e844e2b23/eee9637d-19a4-44c3-bd64-de7e14bae276)

I've confirmed that this trigger isn't working yet:
```
name: Release Trigger
on:
  release:
    types: [published, released]

jobs:
  setup-release:
    runs-on: ubuntu-latest
    outputs:
      git-tag: ${{ steps.tag.outputs.version }}
      env: ${{ steps.setup.outputs.environment }}
    steps:
      - name: Check release
        run: |
          echo "Release tag: ${{ github.event.release.tag_name }}"
          echo "Release name: ${{ github.event.release.name }}"
          echo "Is pre-release (boolean): ${{ github.event.release.prerelease }}"
          echo "Is release (string): ${{ github.event.action }}"
      - name: Setup Environment
        id: setup
        uses: actions/github-script@v6
        with:
          script: |
            if (github.event.release.isPrerelease) {
              core.setOutput('environment', 'staging')
            }
            if ('${{ github.event.action }}' === 'released') {
              core.setOutput('environment', 'prod')
            }
```

**Comment by ldraney** [comment]
Playing with it in my personal repo it seems that the [REST](https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28) documentation is the most useful

**Comment by ldraney** [comment]
My solution looks like this:
```
name: Pre-release Trigger
on:
  release:
    types: [published, released]

jobs:
  setup-release:
    runs-on: ubuntu-latest
    outputs:
      git-tag: ${{ steps.tag.outputs.version }}
      env: ${{ steps.setup.outputs.environment }}
    steps:
      - name: Check release
        run: |
          echo "Release tag: ${{ github.event.release.tag_name }}"
          echo "Release name: ${{ github.event.release.name }}"
          echo "github.event.release.prerelease: ${{ github.event.release.prerelease }}"
          echo "github.event.release.isPrerelease: ${{ github.event.release.isPrerelease }}"
          echo "github.event.action: ${{ github.event.action }}"
          echo "github.payload: ${{ github.event.release.release }}"
      - name: Setup Environment
        id: setup
        uses: actions/github-script@v6
        with:
          script: |
            if ('${{ github.event.release.prerelease }}' === 'true') {
              core.setOutput('environment', 'staging')
            }
            if ('${{ github.event.action }}' === 'released') {
              core.setOutput('environment', 'prod')
            }

  run-deployment:
    needs: [setup-release]
    uses: ./.github/workflows/deployment.yml
    with:
      environment: "${{ needs.setup-release.outputs.env }}"
```
Pros: actions now work as intended
Cons: The workflow runs twice when deploying to prod, though it will effectively only be one, just may confuse the developer who is deploying.

**Comment by ldraney** [comment]
I meant to post the above sooner, sorry about that. 

Kimball, a fellow Oddball engineer, suggested adding:
```
  # A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  init:
    runs-on: ubuntu-latest
    # This filters out pre-release releases
    if: "!github.event.release.prerelease"
    # when a non draft PR is created TODO test actor on new repo!!!!
    steps:
      - name: Init
        run: |
          echo Deploy starting
          echo ${{ github.actor }}
***
  setup-release:
    needs: [init]
```
Based on [how they addressed this problem](https://github.com/department-of-veterans-affairs/chip/blob/master/.github/workflows/prod-deploy.yaml), but it still shows up in our run actions, just grayed out:
![image.png](https://images.zenhubusercontent.com/6361638d438b9c8e844e2b23/73ad0b00-120e-4714-b48d-0559a9280e62)
But at least greyed out is a little better than appearing to both run or one run one fail.  I think I will incorporate this since it does add some clarity to our Actions tab.

**Comment by ldraney** [comment]
I think this is our best solution:
```name: Pre-release Trigger
on:
  release:
    types: [published, released]

  # A workflow run is made up of one or more jobs that can run sequentially or in parallel
jobs:
  init:
    runs-on: ubuntu-latest
    # This filters out pre-release releases
    if: github.event.release.prerelease == true || github.event.action == 'released'
    # when a non draft PR is created TODO test actor on new repo!!!!
    steps:
      - name: Init
        run: |
          echo Deploy starting
          echo ${{ github.actor }}

  setup-release:
    needs: [init]
    runs-on: ubuntu-latest
    outputs:
      git-tag: ${{ steps.tag.outputs.version }}
      env: ${{ steps.setup.outputs.environment }}
    steps:
      - name: Check release
        run: |
          echo "github.event.release.prerelease: ${{ github.event.release.prerelease }}"
          echo "github.event.action: ${{ github.event.action }}"
      - name: Setup Environment
        id: setup
        uses: actions/github-script@v6
        with:
          script: |
            if ('${{ github.event.release.prerelease }}' === true) {
              core.setOutput('environment', 'staging')
            }
            if ('${{ github.event.action }}' === 'released') {
              core.setOutput('environment', 'prod')
            }
            else {
              console.log('Pre-release is false. Stopping workflow.')
              process.exit(0)
            }
```
It works on prerelease as intended with draft. 
and it shows a skipped prerelease trigger like such (ignore that the workflow is called pre-release):
![image.png](https://images.zenhubusercontent.com/6361638d438b9c8e844e2b23/63a28e94-b0ab-416a-925e-9761e788a3af)

**Comment by cris-oddball** [comment]
Lucas created PR, reviewed it with me line by line, we merged it to master, we cut a tag and pushed that tag. Created a pre-release as a draft. Exited the draft. Returned to the draft and published it. The deployment started to staging.

![Screen Shot 2023-06-22 at 11.44.57 AM.jpg](https://images.zenhubusercontent.com/630696b479f5cca51e546872/9af259d6-a2c1-407a-a764-e5bbeffdb945)
